### PR TITLE
2nd Milestone of the React Calculator App

### PR DESCRIPTION
**Pull request by MahmoudBakr23**
# **Hello there**

<hr>

## In this milestone, I've done the following:-

1. I modified the default App STC (Stateless Functional Component) Into (CC) Class Component.
2. I created Display CC which is a child component of the App.
3. I created ButtonPanel CC which is also a child component of the App.
4. I created Button CC which is a child component of the ButtonPanel.
5. I added a ```name``` property to the Button, and a ```result``` property to the Display.
6. I modified a little bit in the ```.eslintrc.json``` file (Just for now) because it's raising issues that don't need to be fixed in this milestone.
7. I merged this milestone to the deploying branch which is (master) you can see the above changes [live](https://mycalc13.herokuapp.com/)

<hr>

Thanks for reviewing.

### CODE REVIEW

**Comment by MahmoudBakr23** [comment]
I've misunderstood the purpose of this milestone. I'm sorry for that!
Please ignore reviewing this PR as proper and mark it as invalid, I'll start doing the required changes by now!
Thank you.

**Comment by MahmoudBakr23** [comment]
Now I hope it is ready for review!
Thanks :)

**Review by Forison — CHANGES_REQUESTED** [review]
## Status::Changes Required🛑

# 👋  Bakr

#### You have done a good job so far, however there are few issues we have to do to get this project approved👨.

> The milestone seeks to introduce you to the container/component pattern, the concept of props the usage and the use of props.

* [x] APP.JS is placed in src/components directory ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/app/App.png))

* [x] to avoid linter error `App` is a functional component ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/app/appFunctionalComponent.png))

> these two requirements seek to make appreciate the concept of container/component pattern, stateless component are referred to as presentation component  hence are placed in the `component` folder,

* [x] App.js has children components wrapped in [React.Fragment](https://reactjs.org/docs/fragments.html#short-syntax) short syntax ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/app/ReactFragment.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/44896092/))

> This requirement seeks to make you appreciate the concept of using fragments in react, you are not suppose to style the component yet, in fact this project does not require styling so we can do the styling when we are done.

* [x] The Display component has `result` prop that should be a String with default value = '0' ([should be String screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/display/validateAsString.png)) ([default props screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/display/DefaultProp.png))

* [x] The Button button name prop should be a String ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/calculator/milestone2/button/validateAsString.png))

* [x] Check comments below👇

#### That is it, kindly make the change and submit a re-review👨‍🎓.

#### Happy coding 👨‍💻

Addo Forison

File: src/App.js [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
-    <div className="App">
-      <h1 className="header">
-        <small>This is just initial.</small>
-        {' '}
-        Hold on!
-      </h1>
-      <code>deployed on Heroku!</code>
-    </div>
-  );
+export default class App extends React.Component {
```
Comment (Forison) [inline src/App.js]: * [x] Change this to functional component

File: src/Components/Button.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import React, { Component } from 'react';
+
+export default class Button extends React.Component {
```
Comment (Forison) [inline src/Components/Button.js]: * [x] Change this to functional component

File: src/Components/ButtonPanel.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import React, { Component } from 'react';
+import Button from './Button';
+
+export default class ButtonPanel extends React.Component {
```
Comment (Forison) [inline src/Components/ButtonPanel.js]: * [x] Change this to functional component

File: src/Components/Display.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import React, { Component } from 'react';
+
+export default class Display extends React.Component {
```
Comment (Forison) [inline src/Components/Display.js]: * [x] Change this to functional component

File: src/Components/Display.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  render() {
+    const { result } = this.props;
+    return (
+      <div className="Display">
+        {result}
+      </div>
+    );
+  }
+}
```
Comment (Forison) [inline src/Components/Display.js]: Add prop validation to track all props that we will use in the component

`
Display.defaultProps = {
  result: '0',
};
`

File: src/Components/Button.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  render() {
+    const { name } = this.props;
+    return (
+      <div className="Button">
+        <button type="button">{name}</button>
+      </div>
+    );
+  }
+}
```
Comment (Forison) [inline src/Components/Button.js]: Add prop validation for each props in a component
Button.propTypes = {name: PropTypes.string.isRequired,}

**Comment by MahmoudBakr23** [comment]
## **Hello Forison**
Thanks so much for reviewing!

<hr>

I did all the changes you asked for, and nothing has changed with the project. It could remains as it was tho!
When I changed to Stateless functional components, it's the same thing to work with classes and I actually prefer classes because it's more organized and simpler.
Regarding the propTypes and copying from a [website](https://developer.fortnox.se/blog/proptypes-in-react-js/#:~:text=PropTypes%20defines%20type%20and%20which,what%20type%20they%20should%20be.)
"PropTypes defines type and which props are required. This benefits the future you and other developer using your component in two ways"
It wasn't necessary and I didn't see them in the requirements, and also my props types were actually a string.
they were wrapped with "prop" like this:
```<Button name="1" />``` That's a string property type. It could be like ```<Button name={1} />```, and that'd be an integer! Both will give the same result but with different prop-types.

File: .eslintrc.json [hunk 7]
```diff
@@ -19,7 +19,12 @@
     "react/react-in-jsx-scope": "off",
     "import/no-unresolved": "off",
     "no-shadow": "off",
-    "arrow-parens": ["error", "as-needed"]
+    "arrow-parens": ["error", "as-needed"],
+    "no-unused-vars": "off",
+    "react/prefer-stateless-function": "off",
+    "no-console": "off",
+    "react/prop-types": "off",
+    "class-methods-use-this": "off"
```
Comment (bellom) [inline .eslintrc.json]: Opps, we have some little changes in this file, check out the [microverse-linter-config](https://github.com/microverseinc/linters-config/blob/master/react-redux/.eslintrc.json)

image
[Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/)

**Review by bellom — CHANGES_REQUESTED** [review]
## Hi @MahmoudBakr23  :clap: 

**Great work making `almost` all the necessary corrections from the previous review.**

# Status: Changes Required! ♻️ 

**Kindly check out my comments below;**

### ♻️ Kindly re-submit for review after making all those changes :slightly_smiling_face: Thumps -up await you.....

**Best,**

**Oluwaseun Bello**
**_Slack Handle - @sean bellom_**

File: .eslintrc.json (hunk 7 above)
Comment (MahmoudBakr23) [reply .eslintrc.json]: I did mention that in the PR summary, I'll change it ok! Thank you.

Comment (MahmoudBakr23) [reply .eslintrc.json]: And you've mentioned that I made 'almost' all the changes, what did I leave?

Comment (bellom) [reply .eslintrc.json]: Just the linter's config file.

**Review by bellom — APPROVED** [review]
## Hi @MahmoudBakr23  :clap: 

**Great work making all the necessary corrections from the previous review.**

# Status: Approved 👍 

## Congratulation :+1:  
_Go ahead and merge to dev and don't forget to submit the project completion form.
Good luck in the next milestone :slightly_smiling_face:_ 

**Best,**

**Oluwaseun Bello**
**_Slack Handle - @sean bellom_**

**Pull request merged**